packages/mattermost-redux/src/reducers/entities/channels.ts

```diff
@@ -516,8 +516,8 @@ function receiveChannelMember(state: RelationOneToOne<Channel, ChannelMembership
             last_viewed_at: Math.max(existingChannelMember.last_viewed_at, received.last_viewed_at),
             msg_count: Math.max(existingChannelMember.msg_count, received.msg_count),
             msg_count_root: Math.max(existingChannelMember.msg_count_root, received.msg_count_root),
-            mention_count: Math.min(existingChannelMember.mention_count, received.mention_count),
-            mention_count_root: Math.min(existingChannelMember.mention_count_root, received.mention_count_root),
```

Comment:
@koox00  @hmhealey I'm not sure why `Math.min` was used in https://github.com/mattermost/mattermost-webapp/pull/10589

If we're using `Math.max` for the message counts, we probably want to use `Math.max` here too. 
With `math.Max` we could end up showing more mention counts, but I think this is better than showing no mention counts with `Math.min`, as is happening in this Jira ticket.



>if (existingChannelMember && updatedChannelMember.last_viewed_at < existingChannelMember.last_viewed_at) {

This code is only executed when this conditional is satisfied (its a few lines above the diff). 

I looked into how the `existingChannelMember` could have an viewed at time that is greater than the newer `updatedChannelMember`

I think this is coming from https://github.com/mattermost/mattermost-webapp/blob/002f6b98309c8c97b099fda0d03bfd7b9dffcf7d/packages/mattermost-redux/src/actions/channels.ts#L789

On channel switch we are setting the last viewed at time to the `now` timestamp. This is incorrect, because on the server we set the last viewed timestamp to the create at of the last post in the channel.

This is why the updatedMembership has a last viewed at before the existing membership. 

Should we fix this in this PR?

Reply:
The reason that it's using max vs min is actually because the naming of these fields is confusing. On ChannelMember objects, `mention_count` is the number of **unread** mentions while `msg_count` is the number of **read** messages, and since I was assuming that a channel that was more recently viewed would always have the number of unreads decreasing, that meant that `mention_count` would go down and `message_count` would go up. That's why `mention_count` used min and `message_count` used max. Oddly enough, I don't know if this change would fix the mention badge for DMs/GMs since I thought that was based on the `message_count`.

---

I think you're right though that this is happening because of the `last_viewed_at` is being set differently on the client than it is on the server. We could try changing that instead to match how the server sets it, but we have to be careful to ensure that there's no existing logic relying on that behaviour on the client.

For example, it used to be the case that you could have the order of the recently viewed channels in the channel switcher not persist through a refresh if you most recently viewed a bunch of already-read channels, but I just tested that now, and that doesn't seem to be the case any more. There's a few other ways that we track the last viewed time for channels (which were workarounds for how the server updates `last_viewed_at` to a post timestamp), so maybe something changed around that logic?

---

Alternatively, maybe we can change the line at the top of `receiveChannelMember` to rely on `last_update_at` instead of `last_viewed_at`? I can't remember why I didn't use that before since I don't think we update that for channel members on the client anywhere. Maybe that'd be more reliable for this?

Reply:
> For example, it used to be the case that you could have the order of the recently viewed channels in the channel switcher not persist through a refresh if you most recently viewed a bunch of already-read channels, but I just tested that now, and that doesn't seem to be the case any more.

Ugh, never mind. This still happens with non-DM/GM channels. DM/GM channels have a user preference they use to keep track of their last viewed time, and it's actually being updated every time they're viewed which I don't think is how they were supposed to work.

Now that I've found that out, changing how `last_viewed_at` is updated on the client will break the sorting of recently viewed channels in the channel switcher

Reply:
Honestly, I think I just forgot that `last_update_at` existed on the `ChannelMember` when I wrote that since `ChannelMember`s don't have the standard CreateAt/DeleteAt/UpdateAt timestamps.

As long as this doesn't cause the race condition from #10589 to occur again, I think changing the if statement should work instead of changing min to max

Reply:
Perfect, thanks for spending some more time on that. Here's hoping we don't have to touch this code for a while after this since the logic is so convoluted 😅